internal/backends/nodejs/nodejs.go

```diff
@@ -48,24 +50,121 @@ type npmSearchResults struct {
 //
 // See https://github.com/npm/registry/blob/5db1bb329f554454467531a3e1bae5e97da160df/docs/responses/package-metadata.md
 // for documentation on the format.
+//
+// ---
+//
+// note: the format isn't 100% validated to the type that's described in the link above.
+// > Historically no validation has been performed on those fields; they are generated by parsing
+// > user-provided data in package.json at publication time.
+//
+// which means that we have to deserialize the union of the type described in the document, and the
+// type of a package.json file: https://docs.npmjs.com/cli/v10/configuring-npm/package-json
 type npmInfoResult struct {
-	Name     string                 `json:"name"`
-	Versions map[string]interface{} `json:"versions"`
-	Author   struct {
-		Name  string `json:"name"`
-		Email string `json:"email"`
-		URL   string `json:"url"`
-	} `json:"author"`
-	Bugs struct {
-		URL string `json:"url"`
-	} `json:"bugs"`
-	Description string `json:"description"`
-	Homepage    string `json:"homepage"`
-	License     string `json:"license"`
-	Repository  struct {
-		Type string `json:"type"`
-		URL  string `json:"url"`
-	} `json:"repository"`
+	Name        string                 `json:"name"`
+	Versions    map[string]interface{} `json:"versions"`
+	Author      packageJsonPerson      `json:"author"`
+	Bugs        packageJsonBugs        `json:"bugs"`
+	Description string                 `json:"description"`
+	Homepage    string                 `json:"homepage"`
+	License     string                 `json:"license"`
+	Repository  packageJsonRepository  `json:"repository"`
+}
+
+type packageJsonRepository struct {
+	URL string
+}
+
+func (repo *packageJsonRepository) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		repo.URL = str
+		return nil
+	}
+
+	var obj map[string]interface{}
+	if err := json.Unmarshal(data, &obj); err == nil {
+		if url, ok := obj["url"].(string); ok {
+			repo.URL = url
+		}
+		return nil
+	}
+
+	return fmt.Errorf("expected repository in package.json to be a string or an object")
+}
+
+type packageJsonPerson struct {
+	Name  string
+	Email string
+	URL   string
+}
+
+var packageJsonPersonStringRegexp *regexp.Regexp
+
+func (person *packageJsonPerson) UnmarshalJSON(data []byte) error {
+	var str string
+	if err := json.Unmarshal(data, &str); err == nil {
+		if packageJsonPersonStringRegexp == nil {
+			packageJsonPersonStringRegexp = regexp.MustCompile(`([^\<\(]*)?\s*(\<.*\>)?\s*(\(.*\))?`)
```

Comment:
Instead of relying on a regex here, could we Unmarshall into a interface{} type and process that? Not sure if it is better, but I think worth exploring if you haven't already.